src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -24,26 +24,79 @@ def self.compile(compile_options)
-          Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
-
-          # This compiler has been configured with a node containing
-          # the requested environment, facts, and variables, and is used
-          # to compile a catalog in that context from the supplied AST.
-          pal.with_catalog_compiler do |compiler|
-            # We have to parse the AST inside the compiler block, because it
-            # initializes the necessary type loaders for us.
-            ast = Puppet::Pops::Serialization::FromDataConverter.convert(code)
-
-            compiler.evaluate(ast)
-            compiler.compile_additions
-            compiler.catalog_data_hash
+                                   facts: compile_options['facts']['values']) do |pal|
+          # No conn info for PE targets can just use empty hash if 
+          # https://github.com/puppetlabs/bolt/pull/1770 is accepted
+          fake_config = {
+            'transport' => 'redacted',
+            'transports' => {
+              'redacted' => 'redacted'
+            }
+          }
+          bolt_inv = Bolt::ApplyInventory.new(fake_config)
+          Puppet.override(bolt_inventory: bolt_inv) do
+            Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
+            #TODO: get rid of this debug logging...
+            Puppet.warning("Compile Options")
+            Puppet.warning("#{compile_options}")
+            # This compiler has been configured with a node containing
+            # the requested environment, facts, and variables, and is used
+            # to compile a catalog in that context from the supplied AST.
+            pal.with_catalog_compiler do |compiler|
+              #CODEREVIEW: Do we need to make these configurable in the request?
+              Puppet[:strict] = :warning
+              Puppet[:strict_variables] = false
+              vars = Puppet::Pops::Serialization::FromDataConverter.convert(compile_options['variables']['values'])
+
+              # If fact conflicts with plan var name, fact wins, so delete plan var
+              # CODEREVIEW: logging is strange for this. For vars that are overridden, that would
+              # need to happen in orch logs, for facts it would happen here. Is it ok to just
+              # silently overwrite?
+              compile_options['facts']['values'].keys.each {|fact_name| vars.delete(fact_name)}
+              Puppet.warning("FACTS KEYS #{compile_options['facts']['values'].keys}")
+              pal.send(:add_variables, compiler.send(:topscope), vars)
+
+              # We have to parse the AST inside the compiler block, because it
+              # initializes the necessary type loaders for us.
+              ast = Puppet::Pops::Serialization::FromDataConverter.convert(code)
+              unless ast.is_a?(Puppet::Pops::Model::Program)
+                Puppet.warning("AST IS A: Puppet::Pops::Model::Program")
```

Comment:
TODO: delete this debug logging

Reply:
also this comment is wrong.. right? since `unless` == `if !`

Reply:
Yeah, this was just frustrated debugging... I was not thinking. It needs to be deleted anyway.

Reply:
Was this going to come out?